### PR TITLE
feat: [PL-43197]: runner watcher

### PR DESCRIPTION
**Pull request by AnupamIO**
Testing:
Tested the changes by executing init task, we are waiting for the pod to be created then proceeding and closing the stream.
Please note that podCreationWait time is kept at 3mins, meaning if pod creation takes >3mins, init request will error out.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56342)

### CODE REVIEW

**Comment by AnupamIO** [comment]
trigger uts

File: 260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java [hunk 1]
```diff
@@ -150,12 +154,13 @@ public void init(
 
       log.info("Creating Task Pod with YAML:\n{}", Yaml.dump(pod));
 
-      // Step 5 - Watch pod logs - normally stop when init finished, but if LE sends response then that's not possible
-      // (e.g. delegate replicaset), but we can stop on watch status
+      coreApi.createNamespacedPod(namespace, pod, null, null, DELEGATE_FIELD_MANAGER, "Warn");
```
Comment (xingchi-jin) [inline 260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java:157]: add a backoff retry for creating the pod. The retry will finish until the timeout is reached 

Later this timeout should come from the UI

**Review by mdmilic — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56342)* status: 6 file(s) to review. 0 of 2 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java` line 81 at r1](https://reviewable.io/reviews/harness/harness-core/56342#-NkzW6JNDqSlsRqI0TLz:-NkzW6JNDqSlsRqI0TM-:bsyfhd5) ([raw file](https://github.com/harness/harness-core/blob/98fcd4450b24754badc33045096822b26367cba7/260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java#L81)):*
> ```Java
>   private final Duration RETRY_SLEEP_DURATION = Duration.ofSeconds(2);
> 
>   private final int POD_INIT_MAX_WAIT_TIME_IN_SEC = 180;
> ```

I think this is good for now, but I believe init timeout is configurable somewhere with 10min default. 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java` line 97 at r1](https://reviewable.io/reviews/harness/harness-core/56342#-NkzWpER9Q9ZZWLvsQIr:-NkzWpER9Q9ZZWLvsQIs:bs65r31) ([raw file](https://github.com/harness/harness-core/blob/98fcd4450b24754badc33045096822b26367cba7/260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java#L97)):*
> ```Java
>   @Override
>   public void init(final String infraId, final InputData infra, final Map<String, char[]> decrypted,
>       final Context context) throws Exception {
> ```

is it possible to define a more specific exception?
___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java` line 68 at r1](https://reviewable.io/reviews/harness/harness-core/56342#-Nl3Hxnm5xiQIgeCYNEk:-Nl3Hxnm5xiQIgeCYNEl:b5m1ee7) ([raw file](https://github.com/harness/harness-core/blob/98fcd4450b24754badc33045096822b26367cba7/260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java#L68)):*
> ```Java
>   }
> 
>   public void waitUntilPodIsReady(CoreV1Api coreV1Api, String podName, String namespace, int podMaxWaitUntilReadySecs)
> ```

Is it possible to get podReady state from a Watch? The same way we log all events from a Watch (like image pulled, container created etc...), I think one of those events should be that containers/pod is ready?
If not that one, maybe creating a Watch on listNamespacedPodCall would give those events?

File: 260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java [hunk 2]
```diff
@@ -42,4 +64,87 @@ public static String getSecretName(final String taskId) {
...
+      List<String> podConditions = pod.getStatus()
+                                       .getConditions()
+                                       .stream()
+                                       .filter(Objects::nonNull)
+                                       .map(V1PodCondition::getMessage)
+                                       .collect(Collectors.toList());
+      errMsg = String.join(", ", podConditions);
+    }
+
+    if (isEmpty(errMsg) && Duration.between(startTime, currTime).getSeconds() >= podMaxWaitUntilReadySecs) {
```
Comment (xingchi-jin) [inline 260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java:119]: this condition` isEmpty(errMsg)`   is not clear to me. what does this mean?
can you create a helper function with proper name  for the function, and wrap the `isEmpty(errMsg) ` in it

**Review by AnupamIO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56342)* status: 7 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java` line 68 at r1](https://reviewable.io/reviews/harness/harness-core/56342#-Nl3Hxnm5xiQIgeCYNEk:-NlP5-Lg9MPd7ru2iA_k:b-qytjk0) ([raw file](https://github.com/harness/harness-core/blob/98fcd4450b24754badc33045096822b26367cba7/260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java#L68)):*
<details><summary><i>Previously, AnupamIO (Anupam Jaiswal) wrote…</i></summary><blockquote>

I'll try with both suggested approaches, had few queries.
1- since we run watcher in async manner, this would mean that we will close watcher in an async manner too when particular message arrives. Is understand correct?
2- If we close watcher in async, what response should we send back? as at this point we don't know if pod will be created or not
</blockquote></details>

Discussed with Marko, both approaches are similar and both will wait until we have a pod status.
1- One polls k8 getPod api every 2s and exits when pod is in ready status. (currently used by CI team for workflows)
2- Create a new watcher which will wait for the create event to happen and after that we will send response back. 

Going ahead with approach1
___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java` line 119 at r1](https://reviewable.io/reviews/harness/harness-core/56342#gh-1421050950:-NlP5lTj1p3hIHgwRr3_:b-2eedp7) ([raw file](https://github.com/harness/harness-core/blob/98fcd4450b24754badc33045096822b26367cba7/260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java#L119)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

this condition` isEmpty(errMsg)`   is not clear to me. what does this mean?
can you create a helper function with proper name  for the function, and wrap the `isEmpty(errMsg) ` in it
</blockquote></details>

Done. Removed time condition as its not needed.

**Review by AnupamIO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56342)* status: 7 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java` line 157 at r1](https://reviewable.io/reviews/harness/harness-core/56342#gh-1418014759:-NlPEHZI4-mBKSFLkbuV:b-j0g2a1) ([raw file](https://github.com/harness/harness-core/blob/98fcd4450b24754badc33045096822b26367cba7/260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java#L157)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

add a backoff retry for creating the pod. The retry will finish until the timeout is reached 

Later this timeout should come from the UI
</blockquote></details>

We have 3 retries for all the k8 client calls with fixed delay of 2s. 
Create call can fail and throw exception when:
`[ApiException](https://javadoc.io/static/io.kubernetes/client-java-api/0.2/io/kubernetes/client/ApiException.html "class in io.kubernetes.client")` \- If fail to call the API, e.g. server error or cannot deserialize the response body

I think if it fails thrice, it should be okay to fail fast and not wait for entire podInitWaitPeriod. So keeping max retries as 3. If you think otherwise, I'll add backoff with max delay as waitTime.

**Comment by AnupamIO** [comment]
trigger uts

**Comment by AnupamIO** [comment]
trigger uts

**Comment by AnupamIO** [comment]
trigger uts

**Review by VikasMaddukuriHarness — APPROVED** [review]

**Pull request merged**